README.md
<p>
  <a href="https://travis-ci.com/Huddly/device-api-ip"><img src="https://travis-ci.com/Huddly/device-api-ip.svg?branch=master" alt="Build Status"></a>
  <a href="https://www.npmjs.com/package/@huddly/device-api-ip"><img src="https://badge.fury.io/js/%40huddly%2Fdevice-api-ip.svg" alt="npm badge"></a>
  <a href="https://npmcharts.com/compare/@huddly/device-api-ip?minimal=true"><img src="https://img.shields.io/npm/dm/@huddly/device-api-ip.svg?style=flat" alt="NPM Downloads"></a>
  <a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/license-MIT-brightgreen.svg" alt="MIT badge"></a>
  <a href="http://commitizen.github.io/cz-cli/"><img src="https://img.shields.io/badge/commitizen-friendly-brightgreen.svg" alt="Commitizen badge"></a>
</p>


# Huddly SDK - Device Api IP
Huddly SDK device api which uses grpc as transport protocol for communicating with the huddly network/ip cameras

## Licence
[MIT](LICENCE)


CHANGELOG.md
### 0.2.1 (2022-08-10)

##### Bug Fixes

*  ws discovery callback only on a timer ([d0d5cb14](https://github.com/Huddly/device-api-ip/commit/d0d5cb14a76894688e1f0aa3b22f7123b7279c6d))
*  there should be detach event for ip device when base is unplugged ([48da7778](https://github.com/Huddly/device-api-ip/commit/48da7778000fad2023b25e65602b9c942747202b))
*  Do not pick up devices with localhost ip ([4e60cbf1](https://github.com/Huddly/device-api-ip/commit/4e60cbf159ac382afacad6289e42b8c21bbfcd37))

#### 0.2.0 (2022-04-04)

##### Chores

*  fix audit issues with minimist ([6e4cc9f8](https://github.com/Huddly/device-api-ip/commit/6e4cc9f8c0f4dbd8d98a2cfdb3735ffd1e3141e0))

##### New Features

*  Update sdk-interfaces to v0.2.0 ([766d6215](https://github.com/Huddly/device-api-ip/commit/766d62159d39cb2e8a679c3a7e401aaaaa9e6803))
*  add s1 as part of the supported devices when getting transport implementation ([ae15f86e](https://github.com/Huddly/device-api-ip/commit/ae15f86ec2197984f77c18d0b7c215b0314aad1a))
*  PID for S1 cameras ([3a3f7319](https://github.com/Huddly/device-api-ip/commit/3a3f73193c2daf5249e5d3497a4cde95321739f7))
* **wsdiscovery:**  Support discovery of S1 Huddly cameras ([80e45adb](https://github.com/Huddly/device-api-ip/commit/80e45adbaf0a8fd140276a238db12ebe3bc10b41))

##### Bug Fixes

* **wsdiscovery:**  message subscriber should not be an anonymous funtion ([f348a1cc](https://github.com/Huddly/device-api-ip/commit/f348a1cc2454a86f92235c0bc208117f8cd35acf))

#### 0.1.6 (2022-02-18)

##### Chores

*  Build & Test with node 16 ([1111ccfa](https://github.com/Huddly/device-api-ip/commit/1111ccfa5e9c808b5cf6cd892d104333420fc3ac))
*  resolve dependeny vulnerabilities ([0812ff2e](https://github.com/Huddly/device-api-ip/commit/0812ff2e1dc8085934051a48f1fcc49354b80ccb))

##### New Features

*  Slack notify when build fails (master) ([debfca33](https://github.com/Huddly/device-api-ip/commit/debfca33d91375e3ea0a8fc3ed26da9a145d4bfe))
*  Cron trigger master branch (Mon-Fri @ 0700) ([7566f332](https://github.com/Huddly/device-api-ip/commit/7566f3326674f97a35f08698596e17781d71f9c0))
*  Allow audit check to have a whitelist ([17b97f8e](https://github.com/Huddly/device-api-ip/commit/17b97f8e9fd062ad76243b7d7083c26d222f2371))
*  Introduce dependency audit-check ([d0579133](https://github.com/Huddly/device-api-ip/commit/d0579133de7ba856eb436580cbd61ade821d6d68))

##### Bug Fixes

*  Wsdd discover only huddly manufactured devices ([08f84221](https://github.com/Huddly/device-api-ip/commit/08f84221ab80fda1508c218a32b4d58882d8f193))
*  Vulnerability check fix with node 16 ([9cb9167b](https://github.com/Huddly/device-api-ip/commit/9cb9167b2d5e5b776df5d3758169990ffa075acc))
*  Update registry url for npmjs to use https ([e2d68e0a](https://github.com/Huddly/device-api-ip/commit/e2d68e0ae39f6b117519d5b0c71048e129198fb3))
* **package.json:**  Win does not recognize '.' ([1187fce3](https://github.com/Huddly/device-api-ip/commit/1187fce32c97c91c69d40741306041452d66ca62))

##### Refactors

*  Remove unused code ([c5d72176](https://github.com/Huddly/device-api-ip/commit/c5d7217670db2aa5ad7ff6f5fdeb191c474b0ca2))

##### Tests

*  Install chalk-js for terminal styling ([9a564e21](https://github.com/Huddly/device-api-ip/commit/9a564e219d443e7bf3fcc8a22c7b8f8c2cff49d7))

#### 0.1.5 (2022-01-26)

##### New Features

*  only probe link local interfaces as default ([3908ae4f](https://github.com/Huddly/device-api-ip/commit/3908ae4fa6cff0da3aae376473bf37303e4d247a))
*  replace travis with githubactions ([a406e921](https://github.com/Huddly/device-api-ip/commit/a406e921cd6d31f6d4d7748785aee25f1b12e3eb))

##### Bug Fixes

* **networkdevice:**  mac addr comparison igore case ([9e2cfda9](https://github.com/Huddly/device-api-ip/commit/9e2cfda981b85d556a509058929d0176ef3a66d8))

##### Other Changes

*  Add ~ for sdk to safly install all patch versions ([31580441](https://github.com/Huddly/device-api-ip/commit/315804419935e59ec906db6c410e68f9f18c1e8f))
*  Emit serialnumber and not device for DETACH event (to correspond with api-usb). ([dda2e0ce](https://github.com/Huddly/device-api-ip/commit/dda2e0cef6d734484468d77bc23725d4d50f3f73))

##### Refactors

*  Use @huddly/sdk-interfaces ([935049ee](https://github.com/Huddly/device-api-ip/commit/935049eee1e66806394876cf51e52311ada3f875))

#### 0.1.4 (2021-08-25)

##### Chores

*  bump sdk dependency to v0.6.1 ([6e49400c](https://github.com/Huddly/device-api-ip/commit/6e49400c647a81d87e0448b286bcf32af540435c))

##### Documentation Changes

* **Readme:**  Add package badges ([98ae5a8c](https://github.com/Huddly/device-api-ip/commit/98ae5a8ceb95005bfe0f8e4a94b48468aaf29cc5))

##### Bug Fixes

*  Update the L1 PID ([9c2f9113](https://github.com/Huddly/device-api-ip/commit/9c2f91136e37b27267127d9803d12fa07525d9ee))



test.js
var Api = require('./lib/src/index').default;
const EventEmitter = require('events');

const main = async () => {

    var api = new Api();
    class MyEmitter extends EventEmitter {}
    const myEmitter = new MyEmitter();
    let grpcTransport;
    let mydevice;


    myEmitter.on('ATTACH', (device) => {
        mydevice = device;
        api.getValidatedTransport(device)
        .then((transport) => {
            grpcTransport = transport;
            grpcClient = transport.grpcClient;
            // TODO: Perform actions
            console.log('We can perform actions on the device now');
            console.log(mydevice);
        }).catch((e) => {
            console.log('Unable to get transport for device');
            console.error(e);
        })
    });
    myEmitter.on('DETACH', (device) => {
        if (mydevice && mydevice.equals(device)) {
            console.log('Closing grpc channel on device');
            grpcTransport.close();
        }
    });
    api.registerForHotplugEvents(myEmitter);
}

main();

tests/api.spec.ts
import chai from 'chai';
import sinonChai from 'sinon-chai';
import sinon from 'sinon';
import { EventEmitter } from 'events';

import GrpcTransport from './../src/transport';
import HuddlyDeviceApiIP from './../src/index';
import HuddlyDevice from './../src/networkdevice';

import HuddlyHEX from '@huddly/sdk-interfaces/lib/enums/HuddlyHex';

const expect = chai.expect;
chai.should();
chai.use(sinonChai);

const mockedDevices: HuddlyDevice[] = [
    new HuddlyDevice({
        name: 'L1',
        pid: HuddlyHEX.L1_PID,
        serialNumber: '1234566445',
        manufacturer: 'Huddly',
        mac: 'FF:FF:FF:FF:FF:FF',
        ip: '',
        types: [],
        scopes: [],
        xaddrs: 'Unknown',
        modelName: 'Unknown',
        metadataVersion: 'Unknown',
        messageId: 'Unknown',
    }),
    new HuddlyDevice({
        name: 'DUMMY',
        serialNumber: '12000045',
        manufacturer: 'JohnDoe Co',
        mac: 'FF:FF:FF:FF:FF:FF',
        ip: '',
        types: [],
        scopes: [],
        xaddrs: 'Unknown',
        modelName: 'Unknown',
        metadataVersion: 'Unknown',
        messageId: 'Unknown',
    }),
    new HuddlyDevice({
        name: 'S1',
        pid: HuddlyHEX.S1_PID,
        serialNumber: '1234566445',
        manufacturer: 'Huddly',
        mac: 'FF:FF:FF:FF:FF:FF',
        ip: '',
        types: [],
        scopes: [],
        xaddrs: 'Unknown',
        modelName: 'Unknown',
        metadataVersion: 'Unknown',
        messageId: 'Unknown',
    }),
];

const dummyDeviceDiscoveryManager = {
    registerForHotplugEvents: () => {},
    deviceList: () => {
        return mockedDevices;
    },
    getDevice: () => {},
};

describe('HuddlyDeviceApiIP', () => {
    let deviceApi: HuddlyDeviceApiIP;
    beforeEach(() => {
        deviceApi = new HuddlyDeviceApiIP({
            manager: dummyDeviceDiscoveryManager,
        });
    });

    describe('#initialize', () => {
        let stub1, stub2, stub3, stub4, stub5, stub6, stub7, stub8;
        afterEach(() => {
            stub1.restore();
            stub2.restore();
            stub3.restore();
            stub4.restore();
            stub5.restore();
            stub6.restore();
            stub7.restore();
            stub8.restore();
        });
        it('should do nothing', () => {
            stub1 = sinon.stub(deviceApi, 'registerForHotplugEvents');
            stub2 = sinon.stub(deviceApi, 'getDeviceDiscoveryAPI');
            stub3 = sinon.stub(deviceApi, 'getValidatedTransport');
            stub4 = sinon.stub(deviceApi, 'getTransport');
            stub5 = sinon.stub(deviceApi, 'isUVCControlsSupported');
            stub6 = sinon.stub(deviceApi, 'getUVCControlAPIForDevice');
            stub7 = sinon.stub(deviceApi, 'isHIDSupported');
            stub8 = sinon.stub(deviceApi, 'getHIDAPIForDevice');
            deviceApi.initialize();
            expect(stub1.called).to.equal(false);
            expect(stub2.called).to.equal(false);
            expect(stub3.called).to.equal(false);
            expect(stub4.called).to.equal(false);
            expect(stub5.called).to.equal(false);
            expect(stub6.called).to.equal(false);
            expect(stub7.called).to.equal(false);
            expect(stub8.called).to.equal(false);
        });
    });
    describe('#registerForHotplugEvents', () => {
        it('should initialize event emitter and register hotplug events on device manager', async () => {
            const emitter = new EventEmitter();
            const spy = sinon.spy(deviceApi.deviceDiscoveryManager, 'registerForHotplugEvents');
            expect(deviceApi.eventEmitter).to.be.undefined;
            await deviceApi.registerForHotplugEvents(emitter);
            expect(deviceApi.eventEmitter).to.be.instanceof(EventEmitter);
            expect(spy.callCount).to.equal(1);
        });
    });

    describe('#getDeviceDiscoveryApi', () => {
        it('should return the device discovery manager instance', async () => {
            const deviceDiscoveryApi = await deviceApi.getDeviceDiscoveryAPI();
            expect(deviceDiscoveryApi).to.equal(dummyDeviceDiscoveryManager);
        });
    });

    describe('#getValidatedTransport', () => {
        describe('for undefined device', () => {
            it('should not initialize a transport implementation', async () => {
                const transport = await deviceApi.getValidatedTransport(undefined);
                expect(transport).to.be.undefined;
            });
        });

        describe('for IP Cameras', () => {
            let transportstub;
            let getTransportStub;
            beforeEach(() => {
                transportstub = sinon.createStubInstance(GrpcTransport);
            });
            afterEach(() => {
                getTransportStub.restore();
            });

            it('should support grpc transport implementation for L1', async () => {
                getTransportStub = sinon.stub(deviceApi, 'getTransport').returns(transportstub);
                const supported = await deviceApi.getValidatedTransport(mockedDevices[0]);
                expect(supported).to.be.instanceof(GrpcTransport);
            });

            it('should support grpc transport implementation for S1', async () => {
                getTransportStub = sinon.stub(deviceApi, 'getTransport').returns(transportstub);
                const supported = await deviceApi.getValidatedTransport(mockedDevices[2]);
                expect(supported).to.be.instanceof(GrpcTransport);
            });

            it('should not support device when init fails', async () => {
                getTransportStub = sinon.stub(deviceApi, 'getTransport').returns(Promise.reject());
                const supported = await deviceApi.getValidatedTransport(mockedDevices[0]);
                expect(supported).to.equal(undefined);
            });

            it('should not support non-Huddly devices', async () => {
                const supported = await deviceApi.getValidatedTransport(mockedDevices[1]);
                expect(supported).to.equal(undefined);
            });
        });
    });

    describe('#getTransport', () => {
        let initStub;
        afterEach(() => {
            initStub.restore();
        });
        it('should initialize GrpcTransport and call init', async () => {
            initStub = sinon.stub(GrpcTransport.prototype, 'init').resolves();
            const supported = await deviceApi.getTransport(mockedDevices[0]);
            expect(initStub.calledOnce).to.equal(true);
            expect(supported).to.be.instanceof(GrpcTransport);
        });
    });

    describe('#isUVCControlsSupported', () => {
        it('should not support UVC controls', async () => {
            const uvcSupport = await deviceApi.isUVCControlsSupported(mockedDevices[0]);
            expect(uvcSupport).to.equal(false);
        });
    });

    describe('#getUVCControlAPIForDevice', () => {
        it('should throw error when calling getUVCControlAPIForDevice for node-usb device api', async () => {
            try {
                await deviceApi.getUVCControlAPIForDevice(mockedDevices[0]);
                expect(true).to.equal(false);
            } catch (e) {
                expect(e.message).to.equal(
                    'UVCControlInterface API not supported for network/ip cameras'
                );
            }
        });
    });

    describe('#isHIDSupported', () => {
        it('should not support HID', async () => {
            const hidSupport = await deviceApi.isHIDSupported(mockedDevices[0]);
            expect(hidSupport).to.equal(false);
        });
    });

    describe('#getHIDApiForDevice', () => {
        it('should throw error when calling getHIDAPIForDevice for node-usb device api', async () => {
            try {
                await deviceApi.getHIDAPIForDevice(mockedDevices[0]);
                expect(true).to.equal(false);
            } catch (e) {
                expect(e.message).to.equal('HID not supported for network/ip cameras');
            }
        });
    });
});


src/index.ts
import { EventEmitter } from 'events';

import IHuddlyDeviceAPI from '@huddly/sdk-interfaces/lib/interfaces/IHuddlyDeviceAPI';
import IUVCControlAPI from '@huddly/sdk-interfaces/lib/interfaces/IUVCControlApi';
import IDeviceDiscovery from '@huddly/sdk-interfaces/lib/interfaces/IDeviceDiscovery';
import DeviceApiOpts from '@huddly/sdk-interfaces/lib/interfaces/IDeviceApiOpts';
import IGrpcTransport from '@huddly/sdk-interfaces/lib/interfaces/IGrpcTransport';
import HuddlyHEX from '@huddly/sdk-interfaces/lib/enums/HuddlyHex';
import Logger from '@huddly/sdk-interfaces/lib/statics/Logger';

import DeviceDiscoveryManager from './manager';
import HuddlyDevice from './networkdevice';
import GrpcTransport from './transport';

export default class HuddlyDeviceApiIP implements IHuddlyDeviceAPI {
    eventEmitter: EventEmitter;
    deviceDiscoveryManager: DeviceDiscoveryManager;

    private readonly SUPPORTED_DEVICE_PIDS: Number[] = [HuddlyHEX.L1_PID, HuddlyHEX.S1_PID];

    constructor(opts: DeviceApiOpts = {}) {
        this.deviceDiscoveryManager = opts.manager || new DeviceDiscoveryManager(opts);
    }

    async initialize() {}

    registerForHotplugEvents(eventEmitter: EventEmitter): void {
        this.eventEmitter = eventEmitter;
        this.deviceDiscoveryManager.registerForHotplugEvents(eventEmitter);
    }

    async getDeviceDiscoveryAPI(): Promise<IDeviceDiscovery> {
        return this.deviceDiscoveryManager;
    }

    async getValidatedTransport(device: HuddlyDevice): Promise<IGrpcTransport> {
        if (!device) {
            Logger.warn('Device is undefined!', HuddlyDeviceApiIP.name);
            return undefined;
        }

        if (!this.SUPPORTED_DEVICE_PIDS.includes(device.productId)) {
            Logger.warn(
                `GRPC is not supported for Huddly device with PID ${device.productId}`,
                HuddlyDeviceApiIP.name
            );
            return undefined;
        }

        try {
            const transport = await this.getTransport(device);
            // TODO: some sort of handshake ?
            Logger.info('Transport protocol is GRPC', HuddlyDeviceApiIP.name);
            return transport;
        } catch (e) {
            // TODO: catch specific exception
            Logger.error(
                `GRPC Transport implementation not supported for device ${device.toString()}`,
                e,
                HuddlyDeviceApiIP.name
            );
            return undefined;
        }
    }

    async getTransport(device: HuddlyDevice): Promise<IGrpcTransport> {
        const transport = new GrpcTransport(device);
        await transport.init();
        return transport;
    }

    async isUVCControlsSupported(_: any) {
        return Promise.resolve(false);
    }

    async getUVCControlAPIForDevice(_: any): Promise<IUVCControlAPI> {
        throw new Error('UVCControlInterface API not supported for network/ip cameras');
    }

    async isHIDSupported(_: any) {
        return Promise.resolve(false);
    }

    async getHIDAPIForDevice(_: any): Promise<any> {
        throw new Error('HID not supported for network/ip cameras');
    }
}


src/transport.ts
import { EventEmitter } from 'events';

import IGrpcTransport from '@huddly/sdk-interfaces/lib/interfaces/IGrpcTransport';
import HuddlyDevice from './networkdevice';

/**
 * Currently depicted as a dummy transport to conform with the structure of the
 * other device-api-* modules. The actual transport will take place on the device
 * manager implementation for the IP cameras.
 *
 * @class GrpcTransport
 * @implements {IGrpcTransport, EventEmitter}
 */
export default class GrpcTransport extends EventEmitter implements IGrpcTransport {
    grpcConnectionDeadlineSeconds: number;
    grpcClient: any;
    device: any;

    constructor(device: HuddlyDevice) {
        super();
        this.device = device;
    }

    overrideGrpcClient(client: any): void {
        this.grpcClient = client;
    }

    init(): Promise<void> {
        return Promise.resolve();
    }
    close(): Promise<any> {
        return Promise.resolve();
    }
}


src/wsdiscovery.ts
import HuddlyDevice from './networkdevice';
import { EventEmitter } from 'events';
import et from 'elementtree';
import { v4 as uuidv4 } from 'node-uuid';
import dgram from 'dgram';
import { networkInterfaces } from 'os';

import Logger from '@huddly/sdk-interfaces/lib/statics/Logger';
import HuddlyHEX from '@huddly/sdk-interfaces/lib/enums/HuddlyHex';

interface InterfaceMap {
    interfaceName: string;
    ip: string;
}

interface DiscoveredInterfaces {
    targetInterface: InterfaceMap;
    linkLocalMaps: InterfaceMap[];
}

export default class WsDiscovery extends EventEmitter {
    maxDelay: number;
    opts: any;
    networkInterfacesWatcher: any;
    socketConnections: NodeJS.Dict<dgram.Socket> = {};

    readonly HUDDLY_MAC_SERIES_START: Number = 0x90e2fc900000;
    readonly HUDDLY_MAC_SERIES_END: Number = 0x90e2fc9fffff;
    readonly HUDDLY_MANUFACTURER_NAME: String = 'Huddly';

    constructor(options: any = {}) {
        super();
        this.opts = options;
        this.opts.timeout = options.timeout || 1000;

        Logger.debug(
            `WsDiscovery initilized with the following options: ${JSON.stringify(this.opts)}`,
            WsDiscovery.name
        );
        this.initNetworkInterfaces();
        this.watchNetworkInterfaces();
    }

    initSocketConnection(
        interfaceMap: InterfaceMap = { interfaceName: 'default', ip: undefined }
    ): dgram.Socket {
        Logger.debug(
            `Discovery probe messages bound to target interface [${interfaceMap.interfaceName}] on addr ${interfaceMap.ip}`,
            WsDiscovery.name
        );
        const socketConnection = dgram.createSocket('udp4');
        this.bindSocket(interfaceMap, socketConnection);
        return socketConnection;
    }

    initNetworkInterfaces(): void {
        const probeEntireNetwork =
            this.opts.probeEntireNetwork || process.env.HUDDLY_WSDD_PROBE_ENTIRE_NETWORK;
        const networkInterfaces = this.findNetworkInterfaces();
        if (this.opts.targetInterfaceAddr || this.opts.targetInterfaceName) {
            const { interfaceName, ip } = networkInterfaces.targetInterface;
            if (!ip) {
                Logger.debug(
                    `Unable to locate target interface with ip address ${this.opts.targetInterfaceAddr} or name ${this.opts.targetInterfaceName}`,
                    WsDiscovery.name
                );
                return;
            }
            this.socketConnections[interfaceName] = this.initSocketConnection(
                networkInterfaces.targetInterface
            );
            return;
        }

        if (probeEntireNetwork) {
            this.socketConnections['default'] = this.initSocketConnection();
            Logger.debug(`Probing entire network for Huddly IP Cameras.`, WsDiscovery.name);
            return;
        }

        if (networkInterfaces.linkLocalMaps.length > 0) {
            networkInterfaces.linkLocalMaps.forEach(map => {
                this.socketConnections[map.interfaceName] = this.initSocketConnection(map);
            });
        } else {
            Logger.debug(`Unable to find any link local networks.`, WsDiscovery.name);
        }
    }

    findNetworkInterfaces(): DiscoveredInterfaces {
        const interfaces = networkInterfaces();
        const targetInterface: InterfaceMap = { interfaceName: undefined, ip: undefined };
        const linkLocalMaps = [];
        for (const [k, v] of Object.entries(interfaces)) {
            if (v instanceof Array) {
                v.forEach((networkInterface: any) => {
                    // Make sure we operate on Ipv4 addresses only
                    if (networkInterface.family === 'IPv4') {
                        // Check if the integrator has specified the target interface or one of the interfaces is huddly compatible (BASE)
                        if (this.opts.targetInterfaceName || this.opts.targetInterfaceAddr) {
                            if (
                                this.opts.targetInterfaceName == k ||
                                this.opts.targetInterfaceAddr == networkInterface.address
                            ) {
                                targetInterface.ip = networkInterface.address;
                                targetInterface.interfaceName = k;
                            }
                        } else if (this.isLinkLocalAddr(networkInterface.address)) {
                            linkLocalMaps.push({
                                ip: networkInterface.address,
                                interfaceName: k,
                            });
                        }
                    }
                });
            }
        }

        return {
            targetInterface,
            linkLocalMaps,
        };
    }

    watchNetworkInterfaces(): void {
        if (this.opts.probeEntireNetwork || process.env.HUDDLY_WSDD_PROBE_ENTIRE_NETWORK) {
            // No need to watch interfaces when socket is bound to all interfaces
            return;
        }

        Logger.debug(`Setting up network interface watcher`, WsDiscovery.name);
        const isNotInitialized = (map: InterfaceMap) => {
            return !(map.interfaceName in this.socketConnections);
        };

        const checkForDisconnect = (maps: InterfaceMap[]) => {
            const flattenedMaps = maps.map(m => m.interfaceName);
            Object.keys(this.socketConnections).forEach(k => {
                if (!flattenedMaps.includes(k)) {
                    Logger.debug(
                        `Network interface [${k}] not found. No longer doing discovery for Huddly Cameras on this interface.`,
                        WsDiscovery.name
                    );

                    this.socketConnections[k].close();
                    this.socketConnections[k] = undefined;
                    delete this.socketConnections[k];
                }
            });
        };

        this.networkInterfacesWatcher = setInterval(() => {
            const networkInterfaces = this.findNetworkInterfaces();
            if (this.opts.targetInterfaceAddr || this.opts.targetInterfaceName) {
                if (
                    networkInterfaces.targetInterface.ip &&
                    isNotInitialized(networkInterfaces.targetInterface)
                ) {
                    this.socketConnections[
                        networkInterfaces.targetInterface.interfaceName
                    ] = this.initSocketConnection(networkInterfaces.targetInterface);
                }
                checkForDisconnect([networkInterfaces.targetInterface]);
                return;
            }

            if (networkInterfaces.linkLocalMaps) {
                networkInterfaces.linkLocalMaps.forEach(map => {
                    if (isNotInitialized(map)) {
                        this.socketConnections[map.interfaceName] = this.initSocketConnection(map);
                    }
                });
            }

            checkForDisconnect(networkInterfaces.linkLocalMaps);
        }, 1000);
    }

    bindSocket(map: InterfaceMap, socket: any): void {
        if (!map.ip) {
            socket.bind();
            return;
        }
        socket.bind({ address: map.ip });
        Logger.debug(
            `Set outgoing multicast interface of socket ${map.interfaceName} to interface with address ${map.ip}`,
            WsDiscovery.name
        );
        socket.on('error', err => {
            this.emit('ERROR', err);
        });
    }

    isLinkLocalAddr(ipAddress: string): boolean {
        return ipAddress.startsWith('169.254');
    }

    isDeviceAllowed(deviceIpAddress: string): boolean {
        // Localhost ip addresses are reserved and not allowed to identify a huddly network camera
        if (deviceIpAddress === '127.0.0.1') return false;

        // This filtering only applies when probing the entire network
        const probeEntireNetwork =
            this.opts.probeEntireNetwork || process.env.HUDDLY_WSDD_PROBE_ENTIRE_NETWORK;
        const ignoreLinkLocalDevices =
            this.opts.ignoreLinkLocalDevices || process.env.HUDDLY_WSDD_IGNORE_LINK_LOCAL_DEVICES;
        const isLinkLocalAddr = this.isLinkLocalAddr(deviceIpAddress);

        if (probeEntireNetwork && ignoreLinkLocalDevices && isLinkLocalAddr) {
            return false;
        }

        return true;
    }

    generateMessageId(): String {
        return 'urn:uuid:' + uuidv4();
    }

    setTimeoutWithRandomDelay(fn: any, max: number): void {
        setTimeout(fn, Math.floor(Math.random() * max));
    }

    manufacturerFromMac(mac: String): String {
        const numericMac = parseInt(mac.split(':').join(''), 16);
        return numericMac >= this.HUDDLY_MAC_SERIES_START &&
            numericMac <= this.HUDDLY_MAC_SERIES_END
            ? this.HUDDLY_MANUFACTURER_NAME
            : '';
    }

    networkDevicePID(name: String): number {
        switch (name) {
            case 'L1':
                return HuddlyHEX.L1_PID;
            case 'S1':
                return HuddlyHEX.S1_PID;
            default:
                return 0x00;
        }
    }

    parseOnvifScopes(scopes: String[], name: String, defaultValue: String[] = ['N/A']): String[] {
        const regex = `(?<=${name}/).*?(?=$|\\s)`;

        const foundValue = [];
        for (let i = 0; i < scopes.length; i++) {
            const element = scopes[i];
            const found = element.match(regex);
            if (found != undefined) {
                foundValue.push(found[0]);
            }
        }
        return foundValue.length > 0 ? foundValue : defaultValue;
    }

    makeDiscoveryBody(msgId: String): Buffer {
        const body =
            '<?xml version="1.0" encoding="UTF-8"?>' +
            '<e:Envelope xmlns:e="http://www.w3.org/2003/05/soap-envelope"' +
            'xmlns:w="http://schemas.xmlsoap.org/ws/2004/08/addressing"' +
            'xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery"' +
            'xmlns:tds="https://www.onvif.org/ver10/device/wsdl/devicemgmt.wsdl"' +
            'xmlns:dn="http://www.onvif.org/ver10/network/wsdl">' +
            '<e:Header>' +
            `<w:MessageID>${msgId}</w:MessageID>` +
            '<w:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</w:To>' +
            '<w:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe</w:Action>' +
            '</e:Header>' +
            '<e:Body>' +
            '<d:Probe><d:Types>dn:NetworkVideoTransmitter</d:Types></d:Probe>' +
            '</e:Body>' +
            '</e:Envelope>';
        return Buffer.from(body);
    }

    probe(callback: any = () => {}): void {
        const messageId = this.generateMessageId();
        const body = this.makeDiscoveryBody(messageId);
        const discoveredDevices: HuddlyDevice[] = [];
        const onProbeResponseHandler = (message: any) => {
            const tree = et.parse(message.toString());
            const relatesTo = tree.findtext('*/wsa:RelatesTo');
            if (relatesTo === messageId) {
                const matches = tree.findall('*/*/wsdd:ProbeMatch');
                matches.forEach(match => {
                    const ipv4Addr = match
                        .findtext('wsdd:XAddrs')
                        .toString()
                        .match(/\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b/);
                    const scopes = match
                        .findtext('wsdd:Scopes')
                        .toString()
                        .split(' ');

                    const macAddr = this.parseOnvifScopes(scopes, 'mac')[0];
                    const name = this.parseOnvifScopes(scopes, 'name', ['Unknown_Device'])[0];
                    const deviceData = {
                        name: name,
                        mac: macAddr,
                        ip: ipv4Addr[0],
                        serialNumber: this.parseOnvifScopes(scopes, 'serial')[0],
                        types: this.parseOnvifScopes(scopes, 'type'),
                        scopes: scopes,
                        xaddrs: match.findtext('wsdd:XAddrs'),
                        modelName: this.parseOnvifScopes(scopes, 'hardware')[0],
                        manufacturer: this.manufacturerFromMac(macAddr),
                        metadataVersion: match.findtext('wsdd:MetadataVersion'),
                        messageId: tree.findtext('*/wsa:MessageID'),
                        pid: this.networkDevicePID(name),
                    };
                    const device = new HuddlyDevice(deviceData);
                    if (this.isDeviceAllowed(device.ip.toString())) {
                        if (device.manufacturer == this.HUDDLY_MANUFACTURER_NAME) {
                            discoveredDevices.push(device);
                            this.emit('device', device);
                        }
                    } else {
                        Logger.debug(
                            `Ignoring device at ip address: ${device.ip.toString()}.`,
                            WsDiscovery.name
                        );
                        Logger.debug(
                            `Ignore local link devices: ${this.opts.ignoreLinkLocalDevices ||
                                process.env.HUDDLY_WSDD_IGNORE_LINK_LOCAL_DEVICES}.`,
                            WsDiscovery.name
                        );
                        Logger.debug(
                            `Has full probing rights: ${this.opts.probeEntireNetwork ||
                                process.env.PROBE_ENTIRE_NETWORK}`,
                            WsDiscovery.name
                        );
                    }
                });
            }
        };

        if (Object.keys(this.socketConnections).length === 0) {
            callback([]);
            return;
        }

        Object.values(this.socketConnections).forEach(socket => {
            socket.on('message', onProbeResponseHandler);

            setTimeout(() => {
                socket.removeListener('message', onProbeResponseHandler);
            }, this.opts.timeout);

            this.setTimeoutWithRandomDelay(
                socket.send.bind(socket, body, 0, body.length, 3702, '239.255.255.250'),
                this.maxDelay
            );
        });
        setTimeout(() => {
            callback(discoveredDevices);
        }, this.opts.timeout);
    }

    close() {
        Object.values(this.socketConnections).forEach(socket => {
            if (socket != undefined) {
                socket.once('close', this.emit.bind(this, 'close'));
                socket.close();
            }
        });
        if (this.networkInterfacesWatcher) {
            clearInterval(this.networkInterfacesWatcher);
        }

        this.socketConnections = undefined;
        this.networkInterfacesWatcher = undefined;
    }
}
